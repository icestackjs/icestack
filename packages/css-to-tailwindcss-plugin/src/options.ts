import path from 'node:path'
import { pathToFileURL } from 'node:url'
import defu from 'defu'
import type { IProcessOptions } from './types'

export function getDefaults(): Required<IProcessOptions> {
  const pkgPath = require.resolve('tailwindcss/package.json')
  const nodeModulesPath = path.dirname(path.dirname(pkgPath))
  return {
    atImportOptions: {},
    sassOptions: {
      importers: [
        {
          // https://github.com/webpack-contrib/sass-loader/blob/master/src/index.js#L55
          findFileUrl(url) {
            const twUrl = pathToFileURL(path.resolve(nodeModulesPath, url))
            return new URL(twUrl)
          }
        }
      ]
    },
    tailwindcssConfig: '',
    generatorOptions: {}
  }
}

export function getOptions(opts?: IProcessOptions): Required<IProcessOptions> {
  return defu(opts, getDefaults())
}
