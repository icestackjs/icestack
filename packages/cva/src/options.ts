import defu from 'defu'
import type { InternalOptions, UserDefinedOptions } from './types'

export function getOptions(opts: UserDefinedOptions = {}) {
  const options: InternalOptions = defu<InternalOptions, InternalOptions[]>(opts, {
    prefix: '',
  })
  return options
}
