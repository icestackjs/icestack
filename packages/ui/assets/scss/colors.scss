$root-vars: (
  --primary-content: #ffffff,
  --primary: #1989fa,
  --success-content: #ffffff,
  --success: #07c160,
  --error-content: #ffffff,
  --error: #ee0a24,
  --warning-content: #ffffff,
  --warning: #ff976a
);

