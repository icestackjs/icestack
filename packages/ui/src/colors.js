const base = require('../assets/js/base/index.js')

function isRgba(colorString) {
  return typeof colorString === 'string' && colorString.includes('/')
}

const colors = {
  transparent: 'transparent',
  current: 'currentColor',
  ...Object.entries(base[':root']).reduce((acc, [key, value]) => {
    // remove -- var prefix
    // "ice-"
    const varName = key.slice('--ice-'.length)
    acc[varName] = isRgba(value) ? `rgba(var(${key}))` : `rgba(var(${key}) / <alpha-value>)`

    return acc
  }, {})
}

module.exports = {
  colors
}
