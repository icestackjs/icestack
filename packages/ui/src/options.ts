import { loadConfig } from 'c12'
import { flattenDeep, get, set, isObject } from 'lodash'
import merge from 'merge'
import { getCodegenDefaults } from './defaults'
import type { CodegenOptions, DeepPartial, Preset } from './types'
import { defuOptions } from '@/shared'
import { makeExtraCssArray } from '@/utils'

export function preHandleOptions(options?: DeepPartial<CodegenOptions>): DeepPartial<CodegenOptions> {
  if (options === undefined) {
    return {}
  }
  if (isObject(options?.base?.themes)) {
    for (const [theme, opts] of Object.entries(options.base?.themes)) {
      if (opts?.extraCss) {
        set(options, `base.themes.${theme}.extraCss`, makeExtraCssArray(opts?.extraCss))
      }
    }
  }

  if (isObject(options?.base) && options.base.extraCss) {
    set(options, `base.extraCss`, makeExtraCssArray(options.base.extraCss))
  }

  if (isObject(options?.utilities) && options.utilities.extraCss && options.utilities.extraCss) {
    set(options, `utilities.extraCss`, makeExtraCssArray(options.utilities.extraCss))
  }

  if (isObject(options?.components)) {
    for (const [componentName, opts] of Object.entries(options.components)) {
      if (opts && isObject(opts.override)) {
        for (const [k, v] of Object.entries(opts.override)) {
          set(options, `components.${componentName}.override.${k}`, makeExtraCssArray(v))
        }
      }
      if (opts && isObject(opts.extend)) {
        for (const [k, v] of Object.entries(opts.extend)) {
          set(options, `components.${componentName}.extend.${k}`, makeExtraCssArray(v))
        }
      }

      if (opts && opts.extra) {
        set(options, `components.${componentName}.extra`, makeExtraCssArray(opts.extra))
      }
    }
  }

  return options
}

export function postHandleOptions(options: DeepPartial<CodegenOptions>): CodegenOptions {
  if (isObject(options?.base?.themes)) {
    for (const [theme, opts] of Object.entries(options.base?.themes)) {
      if (opts?.extraCss) {
        const p = `base.themes.${theme}.extraCss`
        const v = get(options, p)
        if (Array.isArray(v)) {
          set(options, p, merge.recursive(true, ...v))
        }
      }
    }
  }

  if (isObject(options?.base) && options.base.extraCss) {
    const p = `base.extraCss`
    const v = get(options, p)
    if (Array.isArray(v)) {
      set(options, p, merge.recursive(true, ...v))
    }
  }
  if (isObject(options?.utilities) && options.utilities.extraCss && options.utilities.extraCss) {
    const p = `utilities.extraCss`
    const v = get(options, p)
    if (Array.isArray(v)) {
      set(options, p, merge.recursive(true, ...v))
    }
  }

  if (isObject(options?.components)) {
    for (const [componentName, opts] of Object.entries(options.components)) {
      if (opts && isObject(opts.override)) {
        for (const k of Object.keys(opts.override)) {
          const p = `components.${componentName}.override.${k}`
          const v = get(options, p)
          if (Array.isArray(v)) {
            set(options, p, merge.recursive(true, ...v))
          }
        }
      }
      if (opts && isObject(opts.extend)) {
        for (const k of Object.keys(opts.extend)) {
          const p = `components.${componentName}.extend.${k}`
          const v = get(options, p)
          if (Array.isArray(v)) {
            set(options, p, merge.recursive(true, ...v))
          }
        }
      }

      if (opts && opts.extra) {
        const p = `components.${componentName}.extra`
        const v = get(options, p)
        if (Array.isArray(v)) {
          set(options, p, merge.recursive(true, ...v))
        }
      }
    }
  }

  return options as CodegenOptions
}

export function getCodegenOptions(options?: DeepPartial<CodegenOptions>): CodegenOptions {
  let presets: Preset[] = []
  if (options?.presets && Array.isArray(options?.presets)) {
    presets =
      flattenDeep(
        options.presets.map((x) => {
          if (typeof x === 'function') {
            // @ts-ignore
            const res = x()
            return res
          }
          return x
        })
      ).filter(Boolean) ?? []
  }
  const a = preHandleOptions(options)
  const bs = presets.map((p) => {
    return preHandleOptions(p)
  })
  const d = preHandleOptions(getCodegenDefaults(options?.mode))
  const opts = defuOptions<CodegenOptions, DeepPartial<CodegenOptions>[]>(a, ...bs, d)
  const pp = postHandleOptions(opts)
  return pp
}

export async function load(cwd?: string) {
  const { config } = await loadConfig<DeepPartial<CodegenOptions>>({
    name: 'icestack',
    cwd
  })
  return config
}
