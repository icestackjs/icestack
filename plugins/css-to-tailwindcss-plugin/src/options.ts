import type { IProcessOptions, TailwindcssPluginOptions } from './types'
import path from 'node:path'
import process from 'node:process'
import { pathToFileURL } from 'node:url'
import defu from 'defu'

function getDefaultCacheDir() {
  return path.resolve(process.cwd(), 'node_modules', '.css-to-tailwindcss-plugin')
}

export function getDefaults(): IProcessOptions & { cacheDir?: string } {
  const pkgPath = require.resolve('tailwindcss/package.json')
  const nodeModulesPath = path.dirname(path.dirname(pkgPath))
  return {
    atImportOptions: {},
    sassOptions: {
      importers: [
        {
          // https://github.com/webpack-contrib/sass-loader/blob/master/src/index.js#L55
          findFileUrl(url) {
            const twUrl = pathToFileURL(path.resolve(nodeModulesPath, url))
            return new URL(twUrl)
          },
        },
      ],
    },
    tailwindcssResolved: false,
    withOptions: true,
    withOptionsWalkCSSRuleObject(obj) {
      return obj
    },
    cacheDir: getDefaultCacheDir(),
  }
}

export function getOptions(opts?: IProcessOptions) {
  return defu(opts, getDefaults()) as Required<TailwindcssPluginOptions>
}
