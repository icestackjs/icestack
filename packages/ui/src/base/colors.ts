import { generate } from '@ant-design/colors'

export { generate, presetPrimaryColors } from '@ant-design/colors'

export function makeRgbaValue(key: string, slash: boolean = true) {
  return slash ? `rgba(var(${key}) / <alpha-value>)` : `rgba(var(${key}),<alpha-value>)`
}

export const gray: { [key: number]: string } = {
  1: '#ffffff',
  2: '#fafafa',
  3: '#f5f5f5',
  4: '#f0f0f0',
  5: '#d9d9d9',
  6: '#bfbfbf',
  7: '#8c8c8c',
  8: '#595959',
  9: '#434343',
  10: '#262626',
  11: '#1f1f1f',
  12: '#141414',
  13: '#000000'
}

interface ColorGenerateOptions {
  theme?: 'dark' | 'default'
  backgroundColor?: string
}
// type GenerateOptions = Parameters<typeof generate>[1]
export function generateColors(key: string, color: string, dark?: boolean): Record<string, string>
export function generateColors(key: string, color: string, opts: true | ColorGenerateOptions): Record<string, string>
export function generateColors(key: string, color: string, opt?: any) {
  let opts = opt
  if (opts === true) {
    opts = {
      theme: 'dark',
      backgroundColor: '#141414'
    }
  }

  const colors = generate(color, opts)
  return {
    [key]: colors[5],
    [`${key}-hover`]: colors[4],
    [`${key}-active`]: colors[6],
    [`${key}-content`]: gray[1] //  typeof opts === 'object' && opts.theme === 'dark' ? gray[13] : gray[1]
  }
}

// https://github.com/ant-design/ant-design/blob/5393d9ce36821e64590d3f0d07daa0d393a4c299/.dumi/theme/common/Color/ColorStyle.tsx#L13
export const sharedExtraColors = {
  light: {
    'base-100': '#ffffff',
    'base-200': '#f2f2f2',
    'base-300': '#e5e6e6',
    'base-content': '#1f2937',
    ...Object.values(gray).reduce<Record<string, string>>((acc, value, idx) => {
      acc[`ant-neutral-${idx + 1}00`] = value
      return acc
    }, {})
  },
  dark: {
    'base-100': '#1d232a',
    'base-200': '#191e24',
    'base-300': '#15191e',
    'base-content': '#697280',
    ...Object.values(gray).reduce<Record<string, string>>((acc, value, idx) => {
      acc[`ant-neutral-${13 - idx}00`] = value
      return acc
    }, {})
  }
}

export const sharedExtraVars = {
  'rounded-box': '1rem',
  'rounded-btn': '0.5rem',
  'rounded-badge': '1.9rem',
  'animation-btn': '0.25s',
  'animation-input': '0.2s',
  'skeleton-duration': '1.2s',
  'border-btn': '1px',
  'tab-border': '1px',
  'tab-radius': '0.5rem'
}
