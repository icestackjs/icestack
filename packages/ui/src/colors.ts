import { generate } from '@ant-design/colors'

import type { CodegenOptions } from './types'
export { generate, presetPrimaryColors } from '@ant-design/colors'

export function makeRgbaValue(key: string) {
  return `rgba(var(${key}) / <alpha-value>)`
}

export const gray: { [key: number]: string } = {
  1: '#ffffff',
  2: '#fafafa',
  3: '#f5f5f5',
  4: '#f0f0f0',
  5: '#d9d9d9',
  6: '#bfbfbf',
  7: '#8c8c8c',
  8: '#595959',
  9: '#434343',
  10: '#262626',
  11: '#1f1f1f',
  12: '#141414',
  13: '#000000'
}
type GenerateOptions = Parameters<typeof generate>[1]
export function generateColorVars(key: string, color: string, dark?: boolean): Record<string, string>
export function generateColorVars(key: string, color: string, opts: GenerateOptions): Record<string, string>
export function generateColorVars(key: string, color: string, opt?: any) {
  let opts = opt
  if (opts === true) {
    opts = {
      theme: 'dark',
      backgroundColor: '#141414'
    }
  }

  const colors = generate(color, opts)
  return {
    [key]: colors[5],
    [`${key}-hover`]: colors[4],
    [`${key}-active`]: colors[6],
    [`${key}-content`]: gray[1]
  }
}

export function getColors(options: CodegenOptions) {
  const { base, varPrefix: varPrefixOptions } = options
  const { varPrefix } = varPrefixOptions
  return {
    transparent: 'transparent',
    current: 'currentColor',
    ...Object.values(base.types).reduce<Record<string, string>>((acc, cur) => {
      const set = Object.values(cur).reduce<Set<string>>((acc, cur) => {
        for (const x of Object.keys(cur)) {
          acc.add(x)
        }
        return acc
      }, new Set())

      for (const x of set) {
        const key = varPrefix + x
        acc[x] = makeRgbaValue(key)
      }
      return acc
    }, {}),
    ...Object.values(base.themes).reduce<Record<string, string>>((acc, { extraColors }) => {
      if (extraColors) {
        for (const x of Object.keys(extraColors)) {
          const key = varPrefix + x
          acc[x] = makeRgbaValue(key)
        }
      }

      return acc
    }, {})
  }
}
