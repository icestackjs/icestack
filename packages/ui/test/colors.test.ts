import { blue, gray, presetDarkPalettes, presetPalettes } from '@ant-design/colors'
import { getColors } from '@/colors'
import { getCodegenOptions } from '@/options'
describe('colors', () => {
  // it('snap', () => {
  //   expect(colors).toMatchSnapshot()
  // })

  it('getColors case 0', () => {
    expect(getColors(getCodegenOptions())).toMatchSnapshot()
  })

  it('getColors case 1', () => {
    expect(
      getColors(
        getCodegenOptions({
          varPrefix: {
            varPrefix: '--som-'
          }
        })
      )
    ).toMatchSnapshot()
  })

  it('@ant-design/colors', () => {
    expect(blue).toMatchSnapshot('blue')
    expect(gray).toMatchSnapshot('gray')
    expect(presetPalettes).toMatchSnapshot('presetPalettes')
    expect(presetDarkPalettes).toMatchSnapshot('presetDarkPalettes')
  })
})
