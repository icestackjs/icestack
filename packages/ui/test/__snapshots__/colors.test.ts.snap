// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`colors > getColors case 0 1`] = `
{
  "base-100": "rgba(var(--ice-base-100) / <alpha-value>)",
  "base-1000": "rgba(var(--ice-base-1000) / <alpha-value>)",
  "base-1100": "rgba(var(--ice-base-1100) / <alpha-value>)",
  "base-1200": "rgba(var(--ice-base-1200) / <alpha-value>)",
  "base-1300": "rgba(var(--ice-base-1300) / <alpha-value>)",
  "base-200": "rgba(var(--ice-base-200) / <alpha-value>)",
  "base-300": "rgba(var(--ice-base-300) / <alpha-value>)",
  "base-400": "rgba(var(--ice-base-400) / <alpha-value>)",
  "base-500": "rgba(var(--ice-base-500) / <alpha-value>)",
  "base-600": "rgba(var(--ice-base-600) / <alpha-value>)",
  "base-700": "rgba(var(--ice-base-700) / <alpha-value>)",
  "base-800": "rgba(var(--ice-base-800) / <alpha-value>)",
  "base-900": "rgba(var(--ice-base-900) / <alpha-value>)",
  "base-content": "rgba(var(--ice-base-content) / <alpha-value>)",
  "current": "currentColor",
  "error": "rgba(var(--ice-error) / <alpha-value>)",
  "error-active": "rgba(var(--ice-error-active) / <alpha-value>)",
  "error-content": "rgba(var(--ice-error-content) / <alpha-value>)",
  "error-hover": "rgba(var(--ice-error-hover) / <alpha-value>)",
  "neutral": "rgba(var(--ice-neutral) / <alpha-value>)",
  "neutral-active": "rgba(var(--ice-neutral-active) / <alpha-value>)",
  "neutral-content": "rgba(var(--ice-neutral-content) / <alpha-value>)",
  "neutral-hover": "rgba(var(--ice-neutral-hover) / <alpha-value>)",
  "primary": "rgba(var(--ice-primary) / <alpha-value>)",
  "primary-active": "rgba(var(--ice-primary-active) / <alpha-value>)",
  "primary-content": "rgba(var(--ice-primary-content) / <alpha-value>)",
  "primary-hover": "rgba(var(--ice-primary-hover) / <alpha-value>)",
  "success": "rgba(var(--ice-success) / <alpha-value>)",
  "success-active": "rgba(var(--ice-success-active) / <alpha-value>)",
  "success-content": "rgba(var(--ice-success-content) / <alpha-value>)",
  "success-hover": "rgba(var(--ice-success-hover) / <alpha-value>)",
  "transparent": "transparent",
  "warning": "rgba(var(--ice-warning) / <alpha-value>)",
  "warning-active": "rgba(var(--ice-warning-active) / <alpha-value>)",
  "warning-content": "rgba(var(--ice-warning-content) / <alpha-value>)",
  "warning-hover": "rgba(var(--ice-warning-hover) / <alpha-value>)",
}
`;

exports[`colors > getColors case 1 1`] = `
{
  "base-100": "rgba(var(--som-base-100) / <alpha-value>)",
  "base-1000": "rgba(var(--som-base-1000) / <alpha-value>)",
  "base-1100": "rgba(var(--som-base-1100) / <alpha-value>)",
  "base-1200": "rgba(var(--som-base-1200) / <alpha-value>)",
  "base-1300": "rgba(var(--som-base-1300) / <alpha-value>)",
  "base-200": "rgba(var(--som-base-200) / <alpha-value>)",
  "base-300": "rgba(var(--som-base-300) / <alpha-value>)",
  "base-400": "rgba(var(--som-base-400) / <alpha-value>)",
  "base-500": "rgba(var(--som-base-500) / <alpha-value>)",
  "base-600": "rgba(var(--som-base-600) / <alpha-value>)",
  "base-700": "rgba(var(--som-base-700) / <alpha-value>)",
  "base-800": "rgba(var(--som-base-800) / <alpha-value>)",
  "base-900": "rgba(var(--som-base-900) / <alpha-value>)",
  "base-content": "rgba(var(--som-base-content) / <alpha-value>)",
  "current": "currentColor",
  "error": "rgba(var(--som-error) / <alpha-value>)",
  "error-active": "rgba(var(--som-error-active) / <alpha-value>)",
  "error-content": "rgba(var(--som-error-content) / <alpha-value>)",
  "error-hover": "rgba(var(--som-error-hover) / <alpha-value>)",
  "neutral": "rgba(var(--som-neutral) / <alpha-value>)",
  "neutral-active": "rgba(var(--som-neutral-active) / <alpha-value>)",
  "neutral-content": "rgba(var(--som-neutral-content) / <alpha-value>)",
  "neutral-hover": "rgba(var(--som-neutral-hover) / <alpha-value>)",
  "primary": "rgba(var(--som-primary) / <alpha-value>)",
  "primary-active": "rgba(var(--som-primary-active) / <alpha-value>)",
  "primary-content": "rgba(var(--som-primary-content) / <alpha-value>)",
  "primary-hover": "rgba(var(--som-primary-hover) / <alpha-value>)",
  "success": "rgba(var(--som-success) / <alpha-value>)",
  "success-active": "rgba(var(--som-success-active) / <alpha-value>)",
  "success-content": "rgba(var(--som-success-content) / <alpha-value>)",
  "success-hover": "rgba(var(--som-success-hover) / <alpha-value>)",
  "transparent": "transparent",
  "warning": "rgba(var(--som-warning) / <alpha-value>)",
  "warning-active": "rgba(var(--som-warning-active) / <alpha-value>)",
  "warning-content": "rgba(var(--som-warning-content) / <alpha-value>)",
  "warning-hover": "rgba(var(--som-warning-hover) / <alpha-value>)",
}
`;
